source_modelling/community_fault_model.py

```diff
@@ -0,0 +1,551 @@
+                    down_dip_dfc_method=DipMethod(feature["properties"]["Method_Dfc"]),
+                    quality_code=feature["properties"]["QualCode"],
+                    reference=feature["properties"]["References"],
+                    comments=feature["properties"]["Comments"],
+                )
+            )
+    return faults
+
+
+def get_community_fault_model() -> list[CommunityFault]:
+    """Get the community fault model.
+
+    Returns
+    -------
+    list[CommunityFault]
+        A list of CommunityFault objects.
+    """
+    return load_community_fault_model(
+        resources.files(source_modelling) / "NZ_CFM" / "NZ_CFM_v1_0.shp"
+    )
+
+
+def community_fault_model_as_geodataframe() -> gpd.GeoDataFrame:
+    """
+    Convert the community fault model to a GeoDataFrame.
+
+    Returns
+    -------
+    gpd.GeoDataFrame
+        A GeoDataFrame containing the community fault model data, with the
+        coordinate reference system set to EPSG:2193 and the geometry column
+        set to 'trace'. The index is set to the fault names.
+    """
+    model = get_community_fault_model()
+
+    # Transform the trace to WGS84
+    for fault in model:
+        fault.trace = shapely.transform(fault.trace, lambda coord: coord[:, ::-1])
+    return gpd.GeoDataFrame(
+        [vars(fault) for fault in model], geometry="trace", crs="EPSG:2193"
+    ).set_index("name")
+
+
+def parse_timeframe_unit(unit: str) -> Optional[int]:
+    """Parse the timeframe unit from a string.
+
+    Parameters
+    ----------
+    unit : str
+        The unit string.
+
+    Returns
+    -------
+    Optional[int]
+        The parsed unit as an integer, or None if not applicable.
+    """
+    if not unit:
+        return None
+    if "Ma" in unit:
+        return int(1e6)
+
+    match = re.match(r"(([\d,]+)s? )?yrs", unit)
+    if match:
+        return int(match.group(2).replace(",", "")) if match.group(2) else 1
+
+    raise ValueError(f"Unknown time unit {unit}")
+
+
+def fault_sense_parse(sense: str) -> MovementSense:
+    """Parse the fault sense from a string.
+
+    Parameters
+    ----------
+    sense : str
+        The sense string.
+
+    Returns
+    -------
+    MovementSense
+        The parsed MovementSense.
+    """
+    parsed_sense = MovementSense(0)
+    if "dextral" in sense:
+        parsed_sense |= MovementSense.DEXTRAL
+    if "reverse" in sense:
+        parsed_sense |= MovementSense.REVERSE
+    if "sinistral" in sense:
+        parsed_sense |= MovementSense.SINISTRAL
+    if "normal" in sense:
+        parsed_sense |= MovementSense.NORMAL
+    return parsed_sense
+
+
+def most_likely_nodal_plane(
+    faults: list[CommunityFault],
+    centroid: npt.NDArray,
+    nodal_plane_1: NodalPlane,
+    nodal_plane_2: NodalPlane,
+    k_neighbours: int = 5,
+) -> NodalPlane:
+    """Find the most likely nodal plane by a nearest neighbour vote on
+    the strike values of the faults near the centroid.
```

Comment:
Interesting! Is there a reference that describes this idea?

Reply:
This is a pretty standard implementation of a Nearest Neighbour model. Probably doesn't need a good reference but [scikit-learn](https://scikit-learn.org/1.5/modules/neighbors.html) has a description of the different implementations of the model. The only difference is that the nearest neighbour computation is done between faults and points rather than point-to-point. This difference is the only real reason I'm not using scikit-learn for this code.